Automattic/jetpack
# Autoloader: Support unoptimized PSR-4

#### Changes proposed in this Pull Request:

One of the things that the autoloader does when generating the manifest is convert all PSR-4 classes into a classmap. While this is great for performance it can be somewhat error-prone for those who are used to the standard operation of Composer's autoloader. As I've started adding more new classes I've found myself constantly forgetting to dump the autoloader!

This PR proposes utilizing the `-o` option that Composer provides in order to only do this conversion in production builds. We do this by generating a manifest with the namespace paths and versions and then loading via PSR-4 when either a classmap option is unavailable or is of an older version. Once I began working on this change I started to see places that I could refactor to increase test granularity as well as clean up the code a bit. I'll highlight some of the new classes and their purpose as well as the changes I've made to existing classes.

- **Notable Changes**
  - `AutoloadGenerator`
    - Now generates a `jetpack_autoload_psr4.php` manifest containing all of the namespace paths and their corresponding versions when ran using `composer dump-autoload` or `composer install` without the `-o` option.
    - I moved the copied autoloader class files from `vendor/` to `vendor/jetpack-autoloader` so that they are more easily identified. This also would make the rename in #16755 unnecessary! I examined how this would play with existing autoloaders and it would do the same as the referenced PR.
- **New Classes**
  - `Manifest_Handler`
    - After adding a `Psr4_Handler` and copying the `Classes_Handler` I recognized that all of these classes are virtually identical. I then opted to refactor them into a single class and write tests that should cover all three cases.
  - `Version_Loader`
    - With the addition of PSR-4 autoloading and the need for version reconciliation there I felt it would make sense if the autoloading itself took place in its own class.

One scenario I wanted to make sure was adequately handled was if one plugin was using classmap loading while another was using PSR-4 loading. This is a situation that is very present in WooCommerce development and it was important that it be addressed. This is handled in the `Version_Loader` during which it will check both sources and load the class from the newest source.

**_One potentially significant difference highlighted in testing this with WooCommerce is the lack of support for invalid class namespaces when autoloading via PSR-4._** Since no `$autoloadType` is passed to `ClassMapGenerator::createMap()` we've avoiding even [seeing this issue](https://github.com/composer/composer/issues/8403), and if you run an optimized dump, Composer will complain even though our `AutoloadGenerator` will still spit out those files anyway. I'm not completely sure yet how I feel about this one, as again, these classes would not have loaded if Composer was used instead. 

_I think_ users can just move/rename files and they'll work correctly in a backwards-compatible way. My leaning here is that perhaps we should be passing the `$autoloadType` as-expected and be consistent with Composer here. Developers can then note the error when they update the autoloader and fix the classes that are incorrect.

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:

* Set up a WooCommerce development environment cloned to `master`
* Replace the `vendor/automattic/jetpack-autoloader/src` folder with the contents from this PR
* Generate the autoloader using `composer dump-autoloader` and make sure `phpunit tests/php/src` passes
* Generate the autoloader using `composer dump-autoloader -o` and make those test still pass

**(Note that the unoptimized autoloader will give an error but pass the tests, this is because of an invalid class namespace)**

#### Proposed changelog entry for your changes:
* Autoloader: Added support for non-optimized PSR-4 namespace loading.
